File: factorial.cpp
#include <iostream>
#include <omp.h>

long long factorial(int start, int end) {
    if (start > end) {
	    return 1;
    }
    if (start == end) {
	    return start;
    }
    if (end - start == 1) {
	    return start * end;
    }
    int mid = (start + end) / 2;
    long long x, y;

    #pragma omp task shared(x)
    x = factorial(start, mid);

    #pragma omp task shared(y)
    y = factorial(mid + 1, end);

    #pragma omp taskwait
    return x * y;
}

int main() {
    int n = 10;
    long long result = 0;

    #pragma omp parallel
    {
        #pragma omp single
        result = factorial(1, n);
    }

    printf("%lld\n", result);
    return 0;
}

File: factorial.jl
function factorial(start, stop)
    if start > stop
        return 1
    end
    if start == stop
        return start
    end
    if stop - start == 1
        return start * stop
    end

    mid = div(start + stop, 2)
    x = Threads.@spawn factorial(start, mid)
    y = Threads.@spawn factorial(mid + 1, stop)

    return fetch(x) * fetch(y)
end

function main()
    n = 10
    result = factorial(1, n)
    println(result)
end

main()
